### PR TITLE
[4722] [v6] Add skipping enriching urls and push notifications

### LINKED ISSUE #4722: [V6] Add abilities to skip triggering push notification and enriching URLs
### Context & proposal

You can read the context and proposal section in the original ticket seen here: https://github.com/GetStream/stream-chat-android/issues/4687

### Acceptance criteria

To accept this task we need to complete the following items:

- [ ] Implement the same in `v6`

### PR DESCRIPTION
**Pull request by MarinTolic**
### 🎯 Goal

closes #4722

### 🛠 Implementation details

- Add parameters `skipPushNotification` and `skipEnforceUrl` to `Message`
- Expose parameters in all places where the user does not have access to the message, e.g. `MessagesScreen` in  the Compose SDK

### 🧪 Testing

Apply the patches above and in both apps test out the following scenarios:

**Expectation: When `skipPushNotification` and `skipEnforceUrl`are set to true (they are in the patches), URLs should not be enriched  and push notifications should not be triggered.**

1. Send a new message containing a link
2. Send a reply containing a link
3. Send a reply from the gallery containing a link
4. Send a message with a link and attachments
5. From the gallery, delete all attachments from the message in step 4.
6. Update an existing message which contains a link
7. Update an existing non-link containing message with a link
8. Disable internet, send a message with attachments and a link, enable internet
9. Send a message with attachments (easier if there's more of them), disable network mid send,  enable network, long press on the message and click the resend option
10. Repeat the tests for threads

#### Patches

<details>

<summary>Compose patch</summary>

```
Index: stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt b/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt
--- a/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt	(revision 882d298d50c7aa4fc4e87d135c703547106b86f5)
+++ b/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/MessagesActivity.kt	(date 1678790217873)
@@ -50,6 +50,7 @@
 import io.getstream.chat.android.compose.sample.ChatApp
 import io.getstream.chat.android.compose.sample.R
 import io.getstream.chat.android.compose.state.mediagallerypreview.MediaGalleryPreviewResultType
+import io.getstream.chat.android.compose.ui.attachments.StreamAttachmentFactories
 import io.getstream.chat.android.compose.ui.components.composer.MessageInput
 import io.getstream.chat.android.compose.ui.components.messageoptions.defaultMessageOptionsState
 import io.getstream.chat.android.compose.ui.components.reactionpicker.ReactionsPicker
@@ -92,11 +93,16 @@
         super.onCreate(savedInstanceState)
 
         setContent {
-            ChatTheme(dateFormatter = ChatApp.dateFormatter) {
+            ChatTheme(
+                dateFormatter = ChatApp.dateFormatter,
+                attachmentFactories = StreamAttachmentFactories.defaultFactories(skipEnrichUrl = true)
+            ) {
                 MessagesScreen(
                     viewModelFactory = factory,
                     onBackPressed = { finish() },
-                    onHeaderTitleClick = {}
+                    onHeaderTitleClick = {},
+                    skipPushNotification = true,
+                    skipEnrichUrl = true,
                 )
 
                 // MyCustomUi()
```

</details>

<details>

<summary>XML patch</summary>

```
Index: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt
--- a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt	(revision 882d298d50c7aa4fc4e87d135c703547106b86f5)
+++ b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt	(date 1678789256397)
@@ -16,33 +16,55 @@
 
 package io.getstream.chat.ui.sample.feature.chat
 
+import android.content.Context
 import android.os.Bundle
 import android.view.LayoutInflater
 import android.view.View
 import android.view.ViewGroup
 import androidx.activity.OnBackPressedCallback
+import androidx.core.content.ContextCompat
 import androidx.fragment.app.Fragment
 import androidx.fragment.app.viewModels
 import androidx.navigation.fragment.findNavController
 import androidx.navigation.fragment.navArgs
+import io.getstream.chat.android.client.ChatClient
 import io.getstream.chat.android.client.errors.ChatError
 import io.getstream.chat.android.client.utils.Result
-import io.getstream.chat.android.models.Flag
+import io.getstream.chat.android.client.utils.message.isThreadReply
+import io.getstream.chat.android.core.ExperimentalStreamChatApi
+import io.getstream.chat.android.core.internal.InternalStreamChatApi
+import io.getstream.chat.android.models.AttachmentType
+import io.getstream.chat.android.models.ChannelCapabilities
 import io.getstream.chat.android.models.Message
+import io.getstream.chat.android.models.SyncStatus
+import io.getstream.chat.android.models.User
 import io.getstream.chat.android.state.utils.EventObserver
+import io.getstream.chat.android.ui.ChatUI
+import io.getstream.chat.android.ui.common.state.messages.Copy
+import io.getstream.chat.android.ui.common.state.messages.Delete
 import io.getstream.chat.android.ui.common.state.messages.Edit
+import io.getstream.chat.android.ui.common.state.messages.Flag
 import io.getstream.chat.android.ui.common.state.messages.MessageMode
+import io.getstream.chat.android.ui.common.state.messages.Pin
 import io.getstream.chat.android.ui.common.state.messages.Reply
+import io.getstream.chat.android.ui.common.state.messages.Resend
+import io.getstream.chat.android.ui.common.state.messages.ThreadReply
 import io.getstream.chat.android.ui.common.state.messages.list.DeleteMessage
 import io.getstream.chat.android.ui.common.state.messages.list.DeletedMessageVisibility
 import io.getstream.chat.android.ui.common.state.messages.list.EditMessage
 import io.getstream.chat.android.ui.common.state.messages.list.SendAnyway
+import io.getstream.chat.android.ui.common.utils.extensions.isModerationFailed
+import io.getstream.chat.android.ui.feature.messages.list.MessageListViewStyle
+import io.getstream.chat.android.ui.feature.messages.list.options.message.MessageOptionItem
+import io.getstream.chat.android.ui.feature.messages.list.options.message.MessageOptionItemsFactory
 import io.getstream.chat.android.ui.utils.extensions.getCreatedAtOrThrow
+import io.getstream.chat.android.ui.utils.extensions.isGiphyNotEphemeral
 import io.getstream.chat.android.ui.viewmodel.messages.MessageComposerViewModel
 import io.getstream.chat.android.ui.viewmodel.messages.MessageListHeaderViewModel
 import io.getstream.chat.android.ui.viewmodel.messages.MessageListViewModel
 import io.getstream.chat.android.ui.viewmodel.messages.MessageListViewModelFactory
 import io.getstream.chat.android.ui.viewmodel.messages.bindView
+import io.getstream.chat.ui.sample.R
 import io.getstream.chat.ui.sample.common.navigateSafely
 import io.getstream.chat.ui.sample.databinding.FragmentChatBinding
 import io.getstream.chat.ui.sample.feature.common.ConfirmationDialogFragment
@@ -142,7 +164,14 @@
 
     private fun initMessageComposerViewModel() {
         messageComposerViewModel.apply {
-            bindView(binding.messageComposerView, viewLifecycleOwner)
+            bindView(binding.messageComposerView,
+                viewLifecycleOwner,
+                sendMessageButtonClickListener = { message ->
+                    sendMessage(message.apply {
+                        skipEnrichUrl = true
+                        skipPushNotification = true
+                    })
+                })
             messageListViewModel.mode.observe(viewLifecycleOwner) {
                 when (it) {
                     is MessageMode.MessageThread -> {
@@ -156,7 +185,12 @@
                 }
             }
             binding.messageListView.setMessageReplyHandler { _, message ->
-                messageComposerViewModel.performMessageAction(Reply(message))
+                messageComposerViewModel.performMessageAction(Reply(message.apply {
+                    this.apply {
+                        skipEnrichUrl = true
+                        skipPushNotification = true
+                    }
+                }))
             }
             binding.messageListView.setMessageEditHandler { message ->
                 messageComposerViewModel.performMessageAction(Edit(message))
@@ -169,6 +203,7 @@
         }
     }
 
+    @OptIn(InternalStreamChatApi::class, ExperimentalStreamChatApi::class)
     private fun initMessageListViewModel() {
         val calendar = Calendar.getInstance()
         messageListViewModel.apply {
@@ -206,6 +241,29 @@
             }
         }
         binding.messageListView.apply {
+
+            setMessageLongClickListener { message ->
+                childFragmentManager.let { fragmentManager ->
+                    if (message.isModerationFailed(currentUser = ChatClient.instance().getCurrentUser())) {
+                        showModeratedMessageDialog(message)
+                    } else {
+                        val style = requireStyle()
+                        val isEditEnabled = style.editMessageEnabled && !message.isGiphyNotEphemeral()
+                        val viewStyle = style.copy(editMessageEnabled = isEditEnabled)
+
+                        val messageOptionItems = CustomMessageOptionItemsFactory(context).createMessageOptionItems(
+                            selectedMessage = message,
+                            currentUser = ChatUI.currentUserProvider.getCurrentUser(),
+                            isInThread = messageListViewModel.mode.value is MessageMode.MessageThread || message.isThreadReply(),
+                            ownCapabilities = messageListViewModel.ownCapabilities.value ?: setOf(),
+                            style = viewStyle
+                        )
+
+                        showMessageOptionsDialog(fragmentManager, message, messageOptionItems)
+                    }
+                }
+            }
+
             setConfirmDeleteMessageHandler { _, confirmCallback: () -> Unit ->
                 ConfirmationDialogFragment.newDeleteMessageInstance(requireContext())
                     .apply {
@@ -220,18 +278,17 @@
                     }.show(parentFragmentManager, null)
             }
 
-            setFlagMessageResultHandler { result ->
-                if (result is Result.Success || result.isAlreadyExistsError()) {
-                    ConfirmationDialogFragment.newMessageFlaggedInstance(requireContext())
-                        .show(parentFragmentManager, null)
-                }
-            }
-
             setModeratedMessageHandler { message, action ->
                 when (action) {
                     DeleteMessage -> messageListViewModel.onEvent(MessageListViewModel.Event.DeleteMessage(message))
-                    EditMessage -> messageComposerViewModel.performMessageAction(Edit(message))
-                    SendAnyway -> messageListViewModel.onEvent(MessageListViewModel.Event.RetryMessage(message))
+                    EditMessage -> messageComposerViewModel.performMessageAction(Edit(message.apply {
+                        skipEnrichUrl = true
+                        skipPushNotification = true
+                    }))
+                    SendAnyway -> messageListViewModel.onEvent(MessageListViewModel.Event.RetryMessage(message.apply {
+                        skipEnrichUrl = true
+                        skipPushNotification = true
+                    }))
                     else -> Unit
                 }
             }
@@ -257,3 +314,132 @@
         }
     }
 }
+
+public open class CustomMessageOptionItemsFactory(
+    private val context: Context,
+) : MessageOptionItemsFactory {
+
+    /**
+     * Creates [MessageOptionItem]s for the selected message.
+     *
+     * @param selectedMessage The currently selected message.
+     * @param currentUser The currently logged in user.
+     * @param isInThread If the message is being displayed in a thread.
+     * @param ownCapabilities Set of capabilities the user is given for the current channel.
+     * @param style The style to be applied to the view.
+     * @return The list of message option items to display.
+     */
+    @OptIn(InternalStreamChatApi::class)
+    override fun createMessageOptionItems(
+        selectedMessage: Message,
+        currentUser: User?,
+        isInThread: Boolean,
+        ownCapabilities: Set<String>,
+        style: MessageListViewStyle,
+    ): List<MessageOptionItem> {
+        if (selectedMessage.id.isEmpty()) {
+            return emptyList()
+        }
+
+        val selectedMessageUserId = selectedMessage.user.id
+
+        val isTextOnlyMessage = selectedMessage.text.isNotEmpty() && selectedMessage.attachments.isEmpty()
+        val hasLinks =
+            selectedMessage.attachments.any { (it.titleLink != null || it.ogUrl != null) && it.type != AttachmentType.GIPHY }
+        val isOwnMessage = selectedMessageUserId == currentUser?.id
+        val isMessageSynced = selectedMessage.syncStatus == SyncStatus.COMPLETED
+        val isMessageFailed = selectedMessage.syncStatus == SyncStatus.FAILED_PERMANENTLY
+
+        // user capabilities
+        val canQuoteMessage = ownCapabilities.contains(ChannelCapabilities.QUOTE_MESSAGE)
+        val canThreadReply = ownCapabilities.contains(ChannelCapabilities.SEND_REPLY)
+        val canPinMessage = ownCapabilities.contains(ChannelCapabilities.PIN_MESSAGE)
+        val canDeleteOwnMessage = ownCapabilities.contains(ChannelCapabilities.DELETE_OWN_MESSAGE)
+        val canDeleteAnyMessage = ownCapabilities.contains(ChannelCapabilities.DELETE_ANY_MESSAGE)
+        val canEditOwnMessage = ownCapabilities.contains(ChannelCapabilities.UPDATE_OWN_MESSAGE)
+        val canEditAnyMessage = ownCapabilities.contains(ChannelCapabilities.UPDATE_ANY_MESSAGE)
+
+        return listOfNotNull(
+            if (style.retryMessageEnabled && isOwnMessage && isMessageFailed) {
+                MessageOptionItem(
+                    optionText = context.getString(R.string.stream_ui_message_list_resend_message),
+                    optionIcon = ContextCompat.getDrawable(context, style.retryIcon)!!,
+                    messageAction = Resend(selectedMessage.apply {
+                        skipEnrichUrl = true
+                        skipPushNotification = true
+                    }),
+                )
+            } else null,
+            if (style.replyEnabled && isMessageSynced && canQuoteMessage) {
+                MessageOptionItem(
+                    optionText = context.getString(R.string.stream_ui_message_list_reply),
+                    optionIcon = ContextCompat.getDrawable(context, style.replyIcon)!!,
+                    messageAction = Reply(selectedMessage.apply {
+                        skipEnrichUrl = true
+                        skipPushNotification = true
+                    }),
+                )
+            } else null,
+            if (style.threadsEnabled && !isInThread && isMessageSynced && canThreadReply) {
+                MessageOptionItem(
+                    optionText = context.getString(R.string.stream_ui_message_list_thread_reply),
+                    optionIcon = ContextCompat.getDrawable(context, style.threadReplyIcon)!!,
+                    messageAction = ThreadReply(selectedMessage.apply {
+                        skipEnrichUrl = true
+                        skipPushNotification = true
+                    }),
+                )
+            } else null,
+            if (style.copyTextEnabled && (isTextOnlyMessage || hasLinks)) {
+                MessageOptionItem(
+                    optionText = context.getString(R.string.stream_ui_message_list_copy_message),
+                    optionIcon = ContextCompat.getDrawable(context, style.copyIcon)!!,
+                    messageAction = Copy(selectedMessage.apply {
+                        skipEnrichUrl = true
+                        skipPushNotification = true
+                    }),
+                )
+            } else null,
+            if (style.editMessageEnabled && ((isOwnMessage && canEditOwnMessage) || canEditAnyMessage) &&
+                selectedMessage.command != AttachmentType.GIPHY
+            ) {
+                MessageOptionItem(
+                    optionText = context.getString(R.string.stream_ui_message_list_edit_message),
+                    optionIcon = ContextCompat.getDrawable(context, style.editIcon)!!,
+                    messageAction = Edit(selectedMessage.apply {
+                        skipEnrichUrl = true
+                        skipPushNotification = true
+                    }),
+                )
+            } else null,
+            if (style.flagEnabled && !isOwnMessage) {
+                MessageOptionItem(
+                    optionText = context.getString(R.string.stream_ui_message_list_flag_message),
+                    optionIcon = ContextCompat.getDrawable(context, style.flagIcon)!!,
+                    messageAction = Flag(selectedMessage),
+                )
+            } else null,
+            if (style.pinMessageEnabled && isMessageSynced && canPinMessage) {
+                val (pinText, pinIcon) = if (selectedMessage.pinned) {
+                    R.string.stream_ui_message_list_unpin_message to style.unpinIcon
+                } else {
+                    R.string.stream_ui_message_list_pin_message to style.pinIcon
+                }
+
+                MessageOptionItem(
+                    optionText = context.getString(pinText),
+                    optionIcon = ContextCompat.getDrawable(context, pinIcon)!!,
+                    messageAction = Pin(selectedMessage),
+                )
+            } else null,
+            if (style.deleteMessageEnabled && (canDeleteAnyMessage || (isOwnMessage && canDeleteOwnMessage))) {
+                MessageOptionItem(
+                    optionText = context.getString(R.string.stream_ui_message_list_delete_message),
+                    optionIcon = ContextCompat.getDrawable(context, style.deleteIcon)!!,
+                    messageAction = Delete(selectedMessage),
+                    isWarningItem = true,
+                )
+            } else null,
+        )
+    }
+}
```

</details>

#### General
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the ~~`develop`~~ `v5` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

gif

### CODE REVIEW

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/MediaGalleryPreviewActivity.kt [hunk 1]
```diff
@@ -176,11 +177,20 @@ public class MediaGalleryPreviewActivity : AppCompatActivity() {
      * Factory used to build the screen ViewModel given the received message ID.
      */
     private val factory by lazy {
+        val messageId = if (Build.VERSION.SDK_INT >= 33) {
+            intent?.getParcelableExtra(
```
Comment (MarinTolic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/MediaGalleryPreviewActivity.kt:181]: Just updating a now deprecated method.

File: stream-chat-android-compose/api/stream-chat-android-compose.api [hunk 2]
```diff
@@ -367,7 +367,7 @@ public final class io/getstream/chat/android/compose/ui/attachments/content/Link
 }
 
 public final class io/getstream/chat/android/compose/ui/attachments/content/MediaAttachmentContentKt {
```
Comment (MarinTolic) [inline stream-chat-android-compose/api/stream-chat-android-compose.api:369]: We are reflecting changes made in `v5` so I haven't added a changelog here.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/MediaGalleryPreviewActivity.kt [hunk 3]
```diff
@@ -176,11 +177,20 @@ public class MediaGalleryPreviewActivity : AppCompatActivity() {
      * Factory used to build the screen ViewModel given the received message ID.
      */
     private val factory by lazy {
+        val messageId = if (Build.VERSION.SDK_INT >= 33) {
```
Comment (JcMinarro) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/MediaGalleryPreviewActivity.kt:180]:
```suggestion
        val messageId = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
```

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/MediaGalleryPreviewContract.kt [hunk 4]
```diff
@@ -43,6 +43,7 @@ public class MediaGalleryPreviewContract :
             attachmentPosition = input.initialPosition,
             videoThumbnailsEnabled = input.videoThumbnailsEnabled,
             streamCdnImageResizing = input.streamCdnImageResizing,
+            skipEnrichUrl = input.skipEnrichUrl
```
Comment (JcMinarro) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/preview/MediaGalleryPreviewContract.kt:46]:
```suggestion
            skipEnrichUrl = input.skipEnrichUrl,
```

**Review by JcMinarro — APPROVED** [review]
Looks Good.
Some suggestions were added

**Review by TGazica — APPROVED** [review]
Looks good, add tasks for issues discussed offline and update changelog.

### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Compose sample runs & works
- [x] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [x] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

**Comment by MarinTolic** [comment]
> Looks good, add tasks for issues discussed offline.
> ### ballot_box_with_checkReviewer Checklist
> 
>     * [x]  UI Components sample runs & works
> 
>     * [x]  Compose sample runs & works
> 
>     * [x]  UI Changes correct (before & after images)
> 
>     * [ ]  Bugs validated (bugfixes)
> 
>     * [x]  New feature tested and works
> 
>     * [ ]  Release notes and docs clearly describe changes
> 
>     * [x]  All code we touched has new or updated KDocs

The CL was not updated here because this is a `v6` adaptation of a feature we've already published in `v5`, so as far as the users can see, there was only continuity and no new changes.

You can see the original changelog addition here: https://github.com/GetStream/stream-chat-android/pull/4701/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed

**Pull request merged**